compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -395,7 +397,40 @@ fn make_mirror_unadjusted<'a, 'tcx>(
                 .zip(substs.upvar_tys())
                 .map(|(captured_place, ty)| capture_upvar(cx, expr, captured_place, ty))
                 .collect();
-            ExprKind::Closure { closure_id: def_id, substs, upvars, movability }
+
+            // Then introduce fake reads based on what we collected during typeck
+            debug!("RFC2229 = closure_fake_read {:?}", cx.typeck_results().closure_fake_reads);
+            let fake_reads = match cx.typeck_results().closure_fake_reads.get(&def_id) {
+                Some(vals) => vals
+                    .iter()
+                    .filter(|val| match val.base {
+                        HirPlaceBase::Local(_) | HirPlaceBase::Upvar(_) => true,
+                        _ => false,
+                    })
+                    .map(|val| {
+                        let var_hir_id = match val.base {
+                            HirPlaceBase::Local(hir_id) => hir_id,
```

Comment:
We only want to fake read for HirPlace::Upvar

Reply:
Since fake reads were recorded from within the context of the closure they will always be upvar. However since the conversion here is happening in the context of the parent convert_var can figure out if this is ExprRef::Local or ExprRef::UpvarRef